torch/csrc/jit/pickler.cpp

```diff
@@ -234,13 +234,14 @@ void Pickler::pushString(const std::string& string) {
 }
 
 void Pickler::pushGlobal(const std::string& name_temp) {
-  memoized_strings_.push_back(name_temp);
-  auto name = memoized_strings_.back();
-  auto memo_entry = memo_map_.find(&(memoized_strings_.back()));
-  if (memo_entry == memo_map_.end()) {
+  auto memo_entry = memoized_strings_map_.find(name_temp);
+  if (memo_entry == memoized_strings_map_.end()) {
+    memoized_strings_map_.insert({name_temp, memo_id});
     push<OpCode>(OpCode::GLOBAL);
-    pushString(memoized_strings_.back());
-    pushMemoization(&(memoized_strings_.back()));
+    pushString(name_temp);
+
+    // Push BINPUT without adding anything to the memo_map_
+    pushMemoization(nullptr);
```

Comment:
This is a weird API. It decides not to update the memo_map_ in pushMemoization and then carefully uses memo_id here (something secretly updated in pushMemoization) 

Suggestion:

```
memoized_string_map_.insert({name_temp, pushBinPutNext()});
```

`pushBinPutNext()` pushes the BINPUT opcode with an incremented memo_id and then returns it. Then pushMemoization becomes:

```
void pushMemoization(void* item) {
  memo_map_[item] = pushBinPutNext();
}
```